src/runtime-rs/crates/virtcontainers/src/spec_info.rs

```diff
@@ -0,0 +1,82 @@
+// Copyright (c) 2021 Alibaba Cloud
+// Copyright (c) 2021 Ant Group
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+const CRI_CONTAINER_TYPE_KEY_LIST: &[&str] = &[
+    // cri containerd
+    "io.kubernetes.cri.container-type",
+    // cri-o
+    "io.kubernetes.cri-o.ContainerType",
+    // docker shim
+    "io.kubernetes.docker.type",
+];
+
+const CRI_SANDBOX_NAME_KEY_LIST: &[&str] = &[
+    // cri containerd
+    "io.kubernetes.cri.sandbox-id",
+    // cri-o
+    "io.kubernetes.cri-o.SandboxID",
+    // docker shim
+    "io.kubernetes.sandbox.id",
+];
+
+#[derive(Debug, thiserror::Error)]
+pub enum OciSpecInfoError {
+    #[error("cannot infer container type")]
+    UnknownContainerType,
+    #[error("cannot find sandbox id")]
+    MissingSandboxId,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum ContainerType {
+    PodSandbox,
+    PodContainer,
+}
+
+#[derive(Debug, Clone)]
+pub enum ContainerSandboxInfo {
+    Sandbox,
+    Container(String),
+}
+
+pub fn contaier_type(
+    spec: &oci_spec::runtime::Spec,
+) -> std::result::Result<ContainerType, OciSpecInfoError> {
+    let annotations = match spec.annotations() {
+        Some(ann) => ann,
+        None => return Ok(ContainerType::PodSandbox),
+    };
+    for k in CRI_CONTAINER_TYPE_KEY_LIST.iter() {
+        if let Some(type_value) = annotations.get(*k) {
+            match type_value.as_str() {
+                "sandbox" => return Ok(ContainerType::PodSandbox),
```

Comment:
Where the majority of the return values are `Ok(ContainerType::PodSandbox)`, it would probably simplify things to set the default return value at the top in a variable, and for the unique instances change the variable; breaking, and then returning the variable instead of implementing a number of returns.